plugins/org.python.pydev.ast/src/org/python/pydev/ast/codecompletion/revisited/AssignAnalysis.java

```diff
@@ -223,6 +233,36 @@ public TokensList getCompletionsFromDefinition(Definition definition, ICompletio
         return ret;
     }
 
+    private TokensList getCompletionsFromWithDef(ICodeCompletionASTManager manager, ICompletionState state,
+            SourceModule sourceModule,
+            Definition definition) throws CompletionRecursionException {
+        SimpleNode ast = definition.ast;
+        ILocalScope scope = definition.scope;
+        ISimpleNode foundAtASTNode = scope.getFoundAtASTNode();
+        if (ast == null || foundAtASTNode == null) {
+            return new TokensList();
+        }
+        if (ast instanceof Name && foundAtASTNode instanceof With) {
+            With withStmt = (With) foundAtASTNode;
+            for (WithItemType item : withStmt.with_item) {
+                if (item instanceof WithItem) {
+                    WithItem withItem = (WithItem) item;
+                    if (ast.equals(withItem.optional_vars)) {
+                        exprType contextExpr = withItem.context_expr;
+                        ICompletionState copy = state.getCopy();
+                        String token = NodeUtils.getFullRepresentationString(contextExpr);
+                        copy.setCol(contextExpr.beginColumn - 1);
+                        copy.setLine(contextExpr.beginLine - 1);
+                        copy.setActivationToken(token);
+                        copy.pushLookingFor(LookingFor.LOOKING_FOR_WITH_CTX);
```

Comment:
I think we shouldn't do this (it'll affect not only the first scope, but all subsequent ones, so, it's not ideal).

I think this is analogous to when we use the `for` statement, so, I think you should be able to extend `internalGenerateGetCompletionsForModule` in the same place where the `getCompletionsUnpackingForLoop` is called.

So, I think that the actual handling should be close to the for handling in the case where `state.getActivationToken().equals(NodeUtils.getFullRepresentationString(for1.target))` (with the exception that we may need to check for the `__enter__`).